styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -346,76 +348,78 @@ public void cleanup(WorkflowInstance workflowInstance, String executionId) {
   }
 
   @VisibleForTesting
-  void cleanupWithRunState(WorkflowInstance workflowInstance, String executionId) {
-    cleanup(workflowInstance, executionId, pod ->
-        getMainContainerStatus(pod).ifPresent(containerStatus -> {
-          if (hasPullImageError(containerStatus)) {
-            deletePod(workflowInstance, executionId);
-          } else {
-            if (containerStatus.getState().getTerminated() != null) {
-              deletePodIfNonDeletePeriodExpired(workflowInstance, executionId, containerStatus);
-            }
-          }
-        }));
+  void cleanupWithRunState(WorkflowInstance workflowInstance, Pod pod) {
+    final Optional<ContainerStatus> containerStatus = getMainContainerStatus(pod);
+    if (!containerStatus.isPresent()) {
+      deletePod(workflowInstance, pod, "No container status");
+    } else if (isTerminated(containerStatus.get())) {
+      deletePodIfNonDeletePeriodExpired(workflowInstance, pod, containerStatus.get());
+    } else if (hasPullImageError(containerStatus.get())) {
+      deletePod(workflowInstance, pod, "Pull image error");
+    }
   }
 
   @VisibleForTesting
-  void cleanupWithoutRunState(WorkflowInstance workflowInstance, String executionId) {
-    cleanup(workflowInstance, executionId, pod ->
-        getMainContainerStatus(pod).ifPresent(containerStatus -> {
-          if (containerStatus.getState().getTerminated() != null) {
-            deletePodIfNonDeletePeriodExpired(workflowInstance, executionId, containerStatus);
-          } else {
-            // if not terminated, delete it directly
-            deletePod(workflowInstance, executionId);
-          }
-        }));
+  void cleanupWithoutRunState(WorkflowInstance workflowInstance, Pod pod) {
+    final Optional<ContainerStatus> containerStatus = getMainContainerStatus(pod);
+    if (!containerStatus.isPresent()) {
+      deletePod(workflowInstance, pod, "No RunState, no container status");
+    } else if (isTerminated(containerStatus.get())) {
+      deletePodIfNonDeletePeriodExpired(workflowInstance, pod, containerStatus.get());
+    } else {
+      // Refresh the pod if it is not terminated as it might be stale
+      pod = client.pods().withName(pod.getMetadata().getName()).get();
```

Comment:
Let's not assign to the method argument.

Reply:
Usually I would agree but assigning to another variable here incurs the risk of using the old variable by mistake.

Reply:
How about factoring this branch out as a method? It would be easier to follow as well.